parsers/RU.py

```diff
@@ -51,23 +57,66 @@
 
 def fetch_production(zone_key='RU', session=None, target_datetime=None, logger=None) -> list:
     """Requests the last known production mix (in MW) of a given country."""
+    if zone_key == 'RU':
+        # Get data for all zones
+        dfs = {}
+        for subzone_key in ['RU-1', 'RU-2', 'RU-AS']:
+            data = fetch_production(subzone_key, session, target_datetime, logger)
+            df = pd.DataFrame(data).set_index('datetime')
+            dfs[subzone_key] = df['production'].apply(pd.Series).fillna(0)
+
+        # Select the time range
+        min_datetime = max([df.index.min() for df in dfs.values()])
+        if min_datetime > dfs['RU-AS'].index.min():
+            min_datetime = min_datetime - pd.Timedelta(30, 'm')
+        max_datetime = min([df.index.max() for df in dfs.values()])
+        index = pd.date_range(min_datetime, max_datetime, freq='30T', name='datetime')
+        dfs_prod = list(map(lambda df: df.reindex(index).bfill(), dfs.values()))
+
+        # Compute the sum
+        df_prod = reduce(lambda x, y: x+y, dfs_prod)
+
+        # Format to dict
+        df_prod = df_prod.apply(dict, axis=1).reset_index(name='production')
+        df_prod['zoneKey'] = 'RU'
+        df_prod['storage'] = [{} for i in range(len(df_prod))]
+        df_prod['source'] = 'so-ups.ru'
+        data = df_prod.to_dict('records')
+        for row in data:
+            row['datetime'] = row['datetime'].to_pydatetime()
+
+        return data
+
+    elif zone_key == 'RU-1' or zone_key == 'RU-2':
+        return fetch_production_1st_synchronous_zone(zone_key, session, target_datetime)
+    elif zone_key == 'RU-AS':
+        return fetch_production_2nd_synchronous_zone(zone_key, session, target_datetime)
+    else:
+        raise NotImplementedError('This parser is not able to parse given zone')
+
+
+def fetch_production_1st_synchronous_zone(zone_key='RU-1', session=None, target_datetime=None) -> list:
+    zone_key_price_zone_mapper = {
+        'RU-1': 1,
+        'RU-2': 2,
+    }
+    if zone_key not in zone_key_price_zone_mapper:
+        raise NotImplementedError('This parser is not able to parse given zone')
+
     if target_datetime:
-        raise NotImplementedError('This parser is not yet able to parse past dates')
+        target_datetime_tz = arrow.get(target_datetime).to(tz)
+    else:
+        target_datetime_tz = arrow.now(tz)
+    # Query at t gives production from t to t+1
+    # I need to shift 1 to get the last value at t
+    datetime_to_fetch = target_datetime_tz.shift(hours=-1)
+    date = datetime_to_fetch.format('YYYY.MM.DD')
```

Comment:
All dates are now shifted one hour from before. Is this correct?

Reply:
Exactly. Now consumption at hour h means consumption from h-1 to h (or from h-30min to h for East Russia). It seemed more intuitive to me.

Reply:
Actually, you commented here but it doesn't take place here. At this place, I am just saying that if it 00:30 we should fetch the data of yesterday, not today because data cannot exist yet.
The shift takes place on line 143.

Reply:
But my commented code was here, it is a bit misleading, I should move it.

Reply:
Ah, I see. Thanks for explaining.

I know using the end datetime is the standard in other places, but electricityMap uses the start datetime due to the way we standardize data across different frequencies. i.e. a datetime at hour `13:00` mean `13:00` - `14:00` and NOT `12:00-13:00`.  So we would have to shift it back again.